replica/server.go

```diff
@@ -354,25 +379,53 @@ func (s *Server) Restore(url, name string) (err error) {
 		return err
 	}
 
-	if err := backupstore.RestoreDeltaBlockBackup(backupURL, toFile); err != nil {
-		return err
-	}
+	restoreObj := NewRestore(toFile)
 
-	newDisk := disk{
-		Parent:      "",
-		Name:        snapName,
-		Removed:     false,
-		UserCreated: false,
-		Created:     util.Now(),
+	config := &backupstore.DeltaRestoreConfig{
+		BackupURL: backupURL,
+		DeltaOps:  restoreObj,
+		Filename:  toFile,
 	}
 
-	if err := r.encodeToFile(newDisk, snapName+metadataSuffix); err != nil {
+	if err := backupstore.RestoreDeltaBlockBackup(config); err != nil {
+		logrus.Errorf("failed to initiate Restore backup for url[%v]: %v", backupURL, err)
 		return err
 	}
 
-	r, err = r.Revert(snapName, util.Now())
-	if err != nil {
-		return err
-	}
-	return r.Close()
+	periodicChecker := time.NewTicker(ProgressBasedTimeoutInMinute * time.Minute)
+	go func() {
+		for t := range periodicChecker.C {
+			logrus.Debugf("checking restore progress at %v", t)
+			restoreObj.Lock()
+			if restoreObj.RestoreProgress == 100 {
+				restoreObj.Unlock()
+				break
+			}
+			if restoreObj.RestoreError != nil {
+				restoreObj.Unlock()
+				break
+			}
+			now := time.Now()
+			diff := now.Sub(restoreObj.LastUpdatedAt)
+			restoreObj.Unlock()
+
+			if diff.Minutes() > ProgressBasedTimeoutInMinute {
+				logrus.Errorf("no restore update happened since %v minutes. Returning failure",
+					ProgressBasedTimeoutInMinute)
+				periodicChecker.Stop()
+				return
+			}
+		}
+		periodicChecker.Stop()
+
+		if restoreObj.RestoreProgress == 100 {
+			logrus.Infof("Successfully completed backup restore")
+			postRestoreOperations(snapName, r)
```

Comment:
hmm, it's done right here... Not sure why it's done differently in the previous commit.